### PR TITLE
feat: more highlights for the popup menu

### PR DESCRIPTION
**Pull request by mg979**
Schermata 2023-03-06 alle 11 02 17

This is how it looks with the `habamax` color scheme and these highlight groups:
```vim
hi Pmenu          ctermbg=235
hi PmenuSel       ctermfg=234 ctermbg=144
hi PmenuKind      ctermfg=1 ctermbg=235
hi PmenuKindSel   ctermfg=1 ctermbg=144
hi PmenuExtra     ctermfg=243 ctermbg=235 cterm=NONE
hi PmenuExtraSel  ctermfg=0 ctermbg=144 cterm=NONE
```

The file edited is `src/eval.c`, and `omnifunc` is set to `ccomplete#Complete`.

The problem is that at the moment it doesn't have a proper default, so unless you define those groups, they will use the groups for the default color scheme (and it's bad).

Those groups should be linked by default to `Pmenu` and `PmenuSel`, so that there is no difference, unless they are defined in a color scheme.

### CODE REVIEW

**Comment by brammool** [comment]
I suppose it's fine to support the different highlighting.

To link the groups by default, change the highlight_init_both array, instead of highlight_init_light and highlight_init_dark.
Similarly to "default link CursorLineSign SignColumn".

Please add a test.  A screendump test will be the simplest to check that the right attributes are used.

**Comment by habamax** [comment]
Should new hl groups also be added to syntax.txt?

image

**Comment by mg979** [comment]
@habamax yes

Btw do you have any idea on how to test the highlight used inside the popup menu? I was writing a test like
```vim
" Test extra highlight groups for popup menu
func Test_pum_highlights()
  CheckScreendump
  let lines =<< trim END
    setlocal omnifunc=ccomplete#Complete
    call feedkeys("istat\<C-X>\<C-o>", "x")
  END
  call writefile(lines, 'Xscript', 'D')
  let buf = RunVimInTerminal('-S Xscript', {})

  " wait for the script to finish
  call TermWait(buf, 100)

  " Test that extra popup highlight groups are used
  call VerifyScreenDump(buf, 'Test_popup_highlights_01', {})
  call WaitForAssert({-> assert_true('*** WHAT HERE? ***', Screenline(4))})

  call StopVimInTerminal(buf)
endfunc
```

But I don't know how to proceed.

**Comment by habamax** [comment]
Is there a simple way to simulate Pmenu you have on screenshot @mg979 ?
We would need it in vim/colorschemes repo (I guess if test would be added we can grab it from there)?

**Comment by mg979** [comment]
@habamax

Set cwd to `src`.
Source this:
```vim
color habamax
hi PmenuKind      ctermfg=1 ctermbg=235
hi PmenuKindSel   ctermfg=1 ctermbg=144
hi PmenuExtra     ctermfg=243 ctermbg=235 cterm=NONE
hi PmenuExtraSel  ctermfg=0 ctermbg=144 cterm=NONE
```

Then do
```vim
new
setlocal omnifunc=ccomplete#Complete
```

Write something and test omnifunc.

**Comment by mg979** [comment]
@habamax did you manage? Any ideas for the test?

**Comment by habamax** [comment]
@mg979 yes, thank you, I was able to get that Pmenu.

As for test, I've never done screendump tests but the process described in `:h terminal-screendump` and there are examples in test dir, like https://github.com/vim/vim/blob/master/src/testdir/test_popupwin.vim#L3735-L3735

**Comment by mg979** [comment]
Thanks I'll check that.

**Comment by mg979** [comment]
The tests succeed locally but not in CI. There it says `pattern not found` when completing, while locally it finds the matches. In the test I'm using the same `ccomplete#Complete` that should find matches in the `src` directory, but the CI doesn't know about that?

**Comment by habamax** [comment]
I think using `ccomplete#Complete` on a vim src wouldn't be great for a unit test as it is not deterministic (I might be wrong though).

The best thing would be to "construct" simplified Pmenu with all Kind* and Extra* using vimscript and test it. If it is possible.

**Comment by mg979** [comment]
I'll try with that, thanks.

File: src/testdir/test_popup.vim [hunk 1]
```diff
@@ -1250,4 +1250,49 @@ func Test_pum_scrollbar()
...
+  call TermWait(buf)
+  call term_sendkeys(buf, "istat\<C-X>\<C-o>")
+  call TermWait(buf, 50)
+  call VerifyScreenDump(buf, 'Test_pum_highlights_01', {})
+  call term_sendkeys(buf, "\<C-E>\<Esc>u")
+  call TermWait(buf)
+  call StopVimInTerminal(buf)
+endfunc
+
+
```
Comment (zeertzjq) [inline src/testdir/test_popup.vim]: Extra blank line
```suggestion

```

**Comment by mg979** [comment]
@zeertzjq thanks

File: src/testdir/test_popup.vim [hunk 2]
```diff
@@ -1250,4 +1250,70 @@ func Test_pum_scrollbar()
...
+      endif
+      return {
+            \ 'words': [
+            \ { 'word': 'aword1', 'menu': 'extra text 1', 'kind': 'W', },
+            \ { 'word': 'aword2', 'menu': 'extra text 2', 'kind': 'W', },
+            \ { 'word': 'aword3', 'menu': 'extra text 3', 'kind': 'W', },
+            \]}
+    endfunc
+    lcd ../..
+    source runtime/autoload/ccomplete.vim
```
Comment (zeertzjq) [inline src/testdir/test_popup.vim]: No longer needed
```suggestion
```

File: src/testdir/test_popup.vim [hunk 3]
```diff
@@ -1250,4 +1250,70 @@ func Test_pum_scrollbar()
...
```
Comment (zeertzjq) [inline src/testdir/test_popup.vim]: No longer needed
```suggestion
```

**Comment by mg979** [comment]
fixed

**Comment by habamax** [comment]
Have just tried, looks good so far:

image

**Comment by mg979** [comment]
For me it should be ok.

**Pull request closed without merging**